netty/src/main/java/io/grpc/transport/netty/DecoratingHttp2ConnectionEncoder.java

```diff
@@ -0,0 +1,206 @@
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.transport.netty;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http2.Http2Connection;
+import io.netty.handler.codec.http2.Http2ConnectionEncoder;
+import io.netty.handler.codec.http2.Http2Exception;
+import io.netty.handler.codec.http2.Http2Flags;
+import io.netty.handler.codec.http2.Http2FrameWriter;
+import io.netty.handler.codec.http2.Http2Headers;
+import io.netty.handler.codec.http2.Http2LifecycleManager;
+import io.netty.handler.codec.http2.Http2RemoteFlowController;
+import io.netty.handler.codec.http2.Http2Settings;
+
+/**
+ * Delegates all method calls to a concrete {@link Http2ConnectionEncoder} implementation.
+ */
+abstract class DecoratingHttp2ConnectionEncoder implements Http2ConnectionEncoder {
+
+  private final Http2ConnectionEncoder delegate;
+
+  abstract static class Builder implements Http2ConnectionEncoder.Builder {
+
+    protected Http2ConnectionEncoder.Builder delegateBuilder;
+
+    public Builder(Http2ConnectionEncoder.Builder delegateBuilder) {
+      this.delegateBuilder = delegateBuilder;
+    }
+
+    @Override
+    public Builder connection(Http2Connection connection) {
+      delegateBuilder.connection(connection);
+      return this;
+    }
+
+    @Override
+    public Builder lifecycleManager(Http2LifecycleManager lifecycleManager) {
+      delegateBuilder.lifecycleManager(lifecycleManager);
+      return this;
+    }
+
+    @Override
+    public Http2LifecycleManager lifecycleManager() {
+      return delegateBuilder.lifecycleManager();
+    }
+
+    @Override
+    public Builder frameWriter(Http2FrameWriter frameWriter) {
+      delegateBuilder.frameWriter(frameWriter);
+      return this;
+    }
+
+    @Override
+    public Http2ConnectionEncoder build() {
+      return buildInternal(delegateBuilder.build());
```

Comment:
What advantage is there is having buildInternal() vs just having build() work like normal: return delegateBuilder.build() ?